File: libs/workflow_interpretation/src/constraint_product.cpp
// constraint_product.cpp — C++ port of
// paleo_workbench/workflow/interpretation/constraint_product.py (CONV-32).
// See include/pwb/workflow_interpretation/constraint_product.hpp.

#include <pwb/workflow_interpretation/constraint_product.hpp>

#include <pwb/factor_host/canonical_json.hpp>
#include <pwb/workflow_runtime/catalog_seam.hpp>
#include <pwb/workflow_runtime/constraint_versions.hpp>

#include <algorithm>
#include <array>
#include <cctype>
#include <cstdio>
#include <cstdlib>
#include <map>
#include <optional>
#include <string>
#include <utility>
#include <vector>

namespace pwb::workflow_interpretation {
namespace {

// ---- local Python-parity helpers (workflow_runtime/src/python_compat.hpp
// is not part of that library's public surface; the small subset needed here
// is mirrored locally) -----------------------------------------------

std::string strip_ascii(std::string text) {
    const auto is_space = [](unsigned char c) {
        return c == ' ' || c == '\t' || c == '\n' || c == '\r' ||
               c == '\v' || c == '\f';
    };
    std::size_t begin = 0;
    while (begin < text.size() && is_space(text[begin])) ++begin;
    std::size_t end = text.size();
    while (end > begin && is_space(text[end - 1])) --end;
    return text.substr(begin, end - begin);
}

std::string lower_ascii(std::string text) {
    for (char& c : text) {
        c = static_cast<char>(
            std::tolower(static_cast<unsigned char>(c)));
    }
    return text;
}

// Python truthiness of a JSON value.
bool truthy(const Json& value) {
    switch (value.type()) {
    case Json::value_t::null: return false;
    case Json::value_t::boolean: return value.get<bool>();
    case Json::value_t::number_integer:
        return value.get<std::int64_t>() != 0;
    case Json::value_t::number_unsigned:
        return value.get<std::uint64_t>() != 0;
    case Json::value_t::number_float:
        return value.get<double>() != 0.0;
    case Json::value_t::string:
        return !value.get_ref<const std::string&>().empty();
    case Json::value_t::array: return !value.empty();
    case Json::value_t::object: return !value.empty();
    default: return false;
    }
}

// str(x or "") — Python str() of a truthy scalar, "" for falsy/missing.
// (Container values never reach these fields in the host document model.)
std::string str_or(const Json& obj, const char* key) {
    if (!obj.is_object() || !obj.contains(key)) return "";
    const Json& value = obj.at(key);
    if (!truthy(value)) return "";
    return pwb::factor_host::python_str_scalar(value);
}

// Python repr() of a string (single quotes unless the text contains ' and
// not "; minimal control escapes) — used by the frozen CRS messages.
std::string repr_str(const std::string& value) {
    const bool has_single = value.find('\'') != std::string::npos;
    const bool has_double = value.find('"') != std::string::npos;
    const char quote = (has_single && !has_double) ? '"' : '\'';
    std::string out(1, quote);
    for (char c : value) {
        switch (c) {
        case '\\': out += "\\\\"; break;
        case '\n': out += "\\n"; break;
        case '\r': out += "\\r"; break;
        case '\t': out += "\\t"; break;
        default:
            if (c == quote) out += '\\';
            out += c;
        }
    }
    out += quote;
    return out;
}

// Python float(x) for the JSON value forms properties.strength can carry.
std::optional<double> python_float(const Json& value) {
    if (value.is_number()) return value.get<double>();
    if (value.is_boolean()) return value.get<bool>() ? 1.0 : 0.0;
    if (value.is_string()) {
        return pwb::factor_host::python_float_from_string(
            value.get_ref<const std::string&>());
    }
    return std::nullopt;
}

// ---- vocabulary -----------------------------------------------------------

constexpr std::array<const char*, 10> kGeoKinds = {
    "source_direction", "provenance_line", "distribution_line",
    "paleo_shoreline",  "facies_boundary",  "fault",
    "interpolation_boundary", "mask",       "exclusion_area",
    "trend_line",
};

// CONSTRAINT_INTERPOLATION_ROLE (layer_roles.py).
std::map<std::string, std::string> make_role_map() {
    return {
        {"source_direction", "direction"},
        {"trend_line", "direction"},
        {"provenance_line", "direction"},
        {"distribution_line", "direction"},
        {"paleo_shoreline", "boundary"},
        {"facies_boundary", "boundary"},
        {"fault", "break"},
        {"interpolation_boundary", "boundary"},
        {"mask", "boundary"},
        {"exclusion_area", "boundary"},
    };
}

// _ROLE_TO_ENGINE_KIND — engine ConstraintKind VALUES.
std::map<std::string, std::string> make_engine_map() {
    return {
        {"break", "barrier"},
        {"direction", "direction"},
        {"boundary", "boundary_mask"},
    };
}

const std::map<std::string, std::string>& role_map() {
    static const std::map<std::string, std::string> kMap = make_role_map();
    return kMap;
}

const std::map<std::string, std::string>& engine_map() {
    static const std::map<std::string, std::string> kMap = make_engine_map();
    return kMap;
}

// ---- line summary ----------------------------------------------------------

ConstraintLineSummary line_summary(const Json& line) {
    ConstraintLineSummary out;
    out.line_id = str_or(line, "id");
    out.name = str_or(line, "name");
    out.role = str_or(line, "role");

    const Json empty_object = Json::object();
    const Json& props =
        (line.is_object() && line.contains("properties") &&
         !line.at("properties").is_null() && line.at("properties").is_object())
            ? line.at("properties")
            : empty_object;

    std::string geo_kind = str_or(props, "constraint_kind");
    if (!constraint_kind_from_value(geo_kind).has_value()) {
        geo_kind = "";  // unknown geological kind is not guessed
    }
    out.geo_kind = geo_kind;

    out.active = true;  // Python bool(getattr(line, "active", True))
    if (line.is_object() && line.contains("active")) {
        out.active = line.at("active").is_null()
                         ? false  // Python bool(None) == False
                         : truthy(line.at("active"));
    }

    // n_points = len(coordinates) — ALL points, no filtering.
    if (line.is_object() && line.contains("coordinates") &&
        line.at("coordinates").is_array()) {
        out.n_points = static_cast<long long>(line.at("coordinates").size());
    }

    out.strength_declared = props.contains("strength");  // KEY presence
    out.strength = 1.0;
    if (out.strength_declared) {
        const std::optional<double> parsed = python_float(props.at("strength"));
        if (parsed.has_value()) {
            out.strength = *parsed;
        } else {
            out.strength = 1.0;
            out.strength_declared = false;
        }
    }

    out.confidence = str_or(props, "confidence");
    if (out.confidence != "low" && out.confidence != "medium" &&
        out.confidence != "high") {
        out.confidence = "unknown";  // undeclared is never guessed
    }

    out.content_fingerprint = str_or(props, "content_fingerprint");
    return out;
}

}  // namespace

std::optional<std::string> constraint_kind_from_value(const std::string& text) {
    const std::string lowered = lower_ascii(strip_ascii(text));
    if (lowered.empty()) return std::nullopt;
    for (const char* kind : kGeoKinds) {
        if (lowered == kind) return std::string(kind);
    }
    return std::nullopt;
}

std::optional<std::string> interpolation_role_for_kind(const std::string& kind) {
    const auto it = role_map().find(kind);
    if (it == role_map().end()) return std::nullopt;
    return it->second;
}

std::optional<std::string> to_engine_kind(const std::string& geo_kind) {
    const auto resolved = constraint_kind_from_value(geo_kind);
    if (!resolved.has_value()) return std::nullopt;
    const auto role = interpolation_role_for_kind(*resolved);
    if (!role.has_value()) return std::nullopt;
    const auto it = engine_map().find(*role);
    if (it == engine_map().end()) return std::nullopt;
    return it->second;
}

std::string assert_constraints_crs_compatible(
    const std::optional<std::string>& factor_crs,
    const std::optional<std::string>& constraint_crs,
    const std::string& context) {
    const std::string a =
        factor_crs.has_value() ? strip_ascii(*factor_crs) : "";
    const std::string b =
        constraint_crs.has_value() ? strip_ascii(*constraint_crs) : "";
    const std::string prefix = context.empty() ? "" : "[" + context + "] ";
    if (!a.empty() && !b.empty() && a != b) {
        throw ConstraintCrsError(
            prefix + "constraint CRS " + repr_str(b) +
            " differs from factor CRS " + repr_str(a) +
            " — reproject the constraint group before interpolation "
            "(mixed coordinates are never silently mixed)");
    }
    if (!a.empty() && b.empty()) {
        return prefix + "constraint CRS undeclared — interpreted in factor CRS " +
               repr_str(a);
    }
    if (!b.empty() && a.empty()) {
        return prefix + "factor CRS undeclared — constraints assumed " +
               repr_str(b);
    }
    if (a.empty() && b.empty()) {
        return prefix +
               "both CRSs undeclared — coordinates assumed consistent (unverified)";
    }
    return "";
}

Json ConstraintLineSummary::to_dict() const {
    Json out = Json::object();
    out["line_id"] = line_id;
    out["name"] = name;
    out["role"] = role;
    out["geo_kind"] = geo_kind;
    out["active"] = active;
    out["n_points"] = n_points;
    out["strength"] = strength;
    out["strength_declared"] = strength_declared;
    out["confidence"] = confidence;
    out["content_fingerprint"] = content_fingerprint;
    return out;
}

bool ConstraintProduct::has_uncommitted_edits() const {
    return !committed_content_hash.empty() && !content_hash.empty() &&
           committed_content_hash != content_hash;
}

Json ConstraintProduct::to_dict() const {
    Json out = Json::object();
    out["constraint_id"] = constraint_id;
    out["name"] = name;
    out["target_horizon"] = target_horizon;
    Json kinds_arr = Json::array();
    for (const std::string& kind : kinds) kinds_arr.push_back(kind);
    out["kinds"] = std::move(kinds_arr);
    Json engine_arr = Json::array();
    for (const std::string& kind : engine_kinds) engine_arr.push_back(kind);
    out["engine_kinds"] = std::move(engine_arr);
    out["crs"] = crs;
    out["crs_declared"] = crs_declared;
    Json lines_arr = Json::array();
    for (const ConstraintLineSummary& line : lines) {
        lines_arr.push_back(line.to_dict());
    }
    out["lines"] = std::move(lines_arr);
    out["n_active"] = n_active;
    out["committed_version_id"] = committed_version_id;
    out["committed_content_hash"] = committed_content_hash;
    out["content_hash"] = content_hash;
    out["maturity"] = maturity;
    out["staleness"] = staleness;
    out["staleness_detail"] = staleness_detail;
    out["source"] = source;
    out["validity"] = validity;
    return out;
}

ConstraintProduct constraint_product_for_group(
    const Json& document, const Json& group,
    const pwb::workflow_runtime::CatalogRepository* catalog) {
    // The runtime seam's read paths are not const-qualified (Python catalog
    // objects are inherently mutable services); the projection only reads.
    auto* repository = const_cast<pwb::workflow_runtime::CatalogRepository*>(
        catalog);

    std::vector<ConstraintLineSummary> lines;
    if (group.is_object() && group.contains("lines") &&
        group.at("lines").is_array()) {
        for (const Json& line : group.at("lines")) {
            lines.push_back(line_summary(line));
        }
    }

    std::vector<std::string> kinds;
    std::vector<std::string> engine;
    for (const ConstraintLineSummary& line : lines) {
        if (!line.geo_kind.empty() &&
            std::find(kinds.begin(), kinds.end(), line.geo_kind) ==
                kinds.end()) {
            kinds.push_back(line.geo_kind);
        }
        std::optional<std::string> engine_kind;
        if (!line.geo_kind.empty()) {
            engine_kind = to_engine_kind(line.geo_kind);
        }
        if (!line.role.empty()) {
            const auto mapped = engine_map().find(line.role);
            if (mapped != engine_map().end() &&
                std::find(engine.begin(), engine.end(), mapped->second) ==
                    engine.end()) {
                engine.push_back(mapped->second);
            }
        }
        if (engine_kind.has_value() &&
            std::find(engine.begin(), engine.end(), *engine_kind) ==
                engine.end()) {
            engine.push_back(*engine_kind);
        }
    }

    std::string committed_version;
    std::string committed_hash;
    if (repository != nullptr) {
        try {
            const auto latest = pwb::workflow_runtime::
                current_constraint_version(*repository,
                                           str_or(group, "id"));
            if (latest.has_value()) {
                committed_version = latest->version_id;
                // str(meta.get("content_hash") or "")
                const Json& meta = latest->metadata;
                if (meta.is_object() && meta.contains("content_hash") &&
                    truthy(meta.at("content_hash"))) {
                    committed_hash =
                        pwb::factor_host::python_str_scalar(
                            meta.at("content_hash"));
                }
            }
        } catch (...) {  // version chain missing -> honest draft
            committed_version.clear();
            committed_hash.clear();
        }
    }

    std::string content_hash;
    try {
        content_hash =
            pwb::workflow_runtime::constraint_group_content_hash(group).first;
    } catch (...) {  // content hash failure -> "" (projection continues)
        content_hash = "";
    }

    // Group staleness: is the latest commit still the live content?
    // (resolve_constraint_ref vocabulary: current/stale/superseded/unknown;
    // never-committed groups with content -> "uncommitted"; "" = 未评估.)
    std::string staleness;
    std::string staleness_detail;
    if (!committed_version.empty()) {
        try {
            const Json verdict = pwb::workflow_runtime::resolve_constraint_ref(
                document, repository,
                "constraints:" + str_or(group, "id") + ":" + committed_version);
            staleness = str_or(verdict, "status");
            staleness_detail = str_or(verdict, "detail");
        } catch (...) {  // resolution failure -> 未评估
            staleness = "unknown";
            staleness_detail = "约束版本解析失败";
        }
    } else if (!content_hash.empty()) {
        staleness = "uncommitted";
        staleness_detail = "约束组从未提交——无版本链，内容以 live 文档为准";
    }

    std::string validity;
    if (!committed_hash.empty() && !content_hash.empty() &&
        committed_hash != content_hash) {
        validity = "存在未提交编辑（live 内容 ≠ 最新提交）";
    }

    ConstraintProduct out;
    out.constraint_id = str_or(group, "id");
    out.name = str_or(group, "name");
    out.target_horizon = str_or(group, "target_horizon");
    out.kinds = std::move(kinds);
    out.engine_kinds = std::move(engine);
    out.crs = str_or(group, "crs");
    out.crs_declared = !out.crs.empty();
    out.lines = std::move(lines);
    out.n_active = 0;
    for (const ConstraintLineSummary& line : out.lines) {
        if (line.active) ++out.n_active;
    }
    out.committed_version_id = std::move(committed_version);
    out.committed_content_hash = std::move(committed_hash);
    out.content_hash = std::move(content_hash);
    out.maturity = out.committed_version_id.empty() ? "draft" : "committed";
    out.staleness = std::move(staleness);
    out.staleness_detail = std::move(staleness_detail);
    out.validity = std::move(validity);
    return out;
}

std::vector<ConstraintProduct> constraint_products_for_document(
    const Json& document,
    const pwb::workflow_runtime::CatalogRepository* catalog) {
    std::vector<ConstraintProduct> products;
    if (!document.is_object() || !document.contains("constraint_layers") ||
        !document.at("constraint_layers").is_array()) {
        return products;
    }
    for (const Json& group : document.at("constraint_layers")) {
        products.push_back(constraint_product_for_group(document, group, catalog));
    }
    return products;
}

}  // namespace pwb::workflow_interpretation

File: legacy/python_reference/product/paleo_workbench/workflow/interpretation/constraint_product.py
"""ConstraintProduct — 约束组领域投影 + 双枚举收敛 + CRS 纪律（V9 ADR-6）。

一个约束组的身份 = ``ConstraintLayers.id``（与 constraint_versions 的
提交/版本链一致）。本模块收敛：

* **双 ConstraintKind 收敛**：地质语义 ``layer_roles.ConstraintKind``
  （10 态）是规范类型；:func:`to_engine_kind` 经
  ``CONSTRAINT_INTERPOLATION_ROLE`` 唯一映射到 engine 语义
  ``constraint_capabilities.ConstraintKind``（5 态）——同名异义枚举不再
  直传（基线 P1-4）；
* **CRS 纪律**：:func:`assert_constraints_crs_compatible` —— 已声明且
  不同的 CRS → ValueError（此前从不比对，基线 P1-3）；未声明 → 诚实
  note（按任务 CRS 解释），绝不猜测；
* **strength/confidence**：per-line ``properties.strength/confidence``
  可选；缺省 strength=1.0（标注 default）confidence=unknown（goal §10）。

编辑生命周期（goal §11）：QGIS 负责几何编辑；commit（既有
``commit_constraint_group``）产生新 DataVersion；下游 factor 经
content-hash pins 感知（本投影只读）。
"""
from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any

from paleo_workbench.mapping_workspace.layer_roles import (
    CONSTRAINT_INTERPOLATION_ROLE,
    ConstraintKind as GeoConstraintKind,
    constraint_kind_from_value,
)
from paleo_workbench.workflow.constraint_capabilities import (
    ConstraintKind as EngineConstraintKind,
)

#: ConstraintLine.role（插值引擎兼容语义）→ engine 约束词汇。
_ROLE_TO_ENGINE_KIND: dict[str, EngineConstraintKind] = {
    "break": EngineConstraintKind.BARRIER,
    "direction": EngineConstraintKind.DIRECTION,
    "boundary": EngineConstraintKind.BOUNDARY_MASK,
}


def to_engine_kind(geo_kind: "GeoConstraintKind | str") -> EngineConstraintKind | None:
    """地质约束类型 → engine 约束类型（唯一映射；未知 → None，不猜）。"""
    resolved = (constraint_kind_from_value(str(geo_kind))
                if not isinstance(geo_kind, GeoConstraintKind) else geo_kind)
    if resolved is None:
        return None
    role = CONSTRAINT_INTERPOLATION_ROLE.get(resolved, "")
    return _ROLE_TO_ENGINE_KIND.get(str(role))


def assert_constraints_crs_compatible(
    factor_crs: str | None,
    constraint_crs: str | None,
    *,
    context: str = "",
) -> str:
    """约束 CRS 与 factor CRS 的纪律检查（基线 P1-3）。

    * 双方已声明且不同 → ``ValueError``（绝不静默混合坐标）；
    * 任一方未声明 → 返回诚实 note（未声明方按对方坐标系解释）；
    * 双方一致 → 返回 ""（无异常说明）。
    """
    a = str(factor_crs or "").strip()
    b = str(constraint_crs or "").strip()
    prefix = f"[{context}] " if context else ""
    if a and b and a != b:
        raise ValueError(
            f"{prefix}constraint CRS {b!r} differs from factor CRS {a!r} — "
            "reproject the constraint group before interpolation "
            "(mixed coordinates are never silently mixed)"
        )
    if a and not b:
        return f"{prefix}constraint CRS undeclared — interpreted in factor CRS {a!r}"
    if b and not a:
        return f"{prefix}factor CRS undeclared — constraints assumed {b!r}"
    if not a and not b:
        return f"{prefix}both CRSs undeclared — coordinates assumed consistent (unverified)"
    return ""


@dataclass(frozen=True)
class ConstraintLineSummary:
    """单条约束线的投影（identity/role/kind/strength/confidence/geometry）。"""

    line_id: str
    name: str
    role: str                       # engine role (break/direction/boundary/...)
    geo_kind: str                   # 地质类型（""=未标注）
    active: bool
    n_points: int
    strength: float = 1.0
    strength_declared: bool = False
    confidence: str = "unknown"     # unknown | low | medium | high（声明才有效）
    content_fingerprint: str = ""

    def to_dict(self) -> dict[str, Any]:
        return {
            "line_id": self.line_id,
            "name": self.name,
            "role": self.role,
            "geo_kind": self.geo_kind,
            "active": self.active,
            "n_points": self.n_points,
            "strength": self.strength,
            "strength_declared": self.strength_declared,
            "confidence": self.confidence,
            "content_fingerprint": self.content_fingerprint,
        }


@dataclass(frozen=True)
class ConstraintProduct:
    """一个约束组的产品投影（载荷=文档 ConstraintLayers；版本=catalog）。"""

    constraint_id: str              # = ConstraintLayers.id（提交/版本链身份）
    name: str = ""
    target_horizon: str = ""
    kinds: tuple[str, ...] = ()     # 地质类型（去重）
    engine_kinds: tuple[str, ...] = ()  # engine 投影（去重）
    crs: str = ""
    crs_declared: bool = False
    lines: tuple[ConstraintLineSummary, ...] = ()
    n_active: int = 0
    #: 最新提交（"" = 从未提交 → maturity=draft）。
    committed_version_id: str = ""
    committed_content_hash: str = ""
    content_hash: str = ""          # 当前内容（含未提交编辑）
    maturity: str = "draft"         # draft | committed
    #: constraint_pins_staleness 词汇：current/unpinned/unknown/missing/
    #: stale_version/stale_content（""=未评估）。
    staleness: str = ""
    staleness_detail: str = ""
    source: str = "user"            # user | imported | derived
    validity: str = ""              # 诚实校验说明（空=未见异常）

    @property
    def has_uncommitted_edits(self) -> bool:
        return bool(
            self.committed_content_hash
            and self.content_hash
            and self.committed_content_hash != self.content_hash
        )

    def to_dict(self) -> dict[str, Any]:
        return {
            "constraint_id": self.constraint_id,
            "name": self.name,
            "target_horizon": self.target_horizon,
            "kinds": list(self.kinds),
            "engine_kinds": list(self.engine_kinds),
            "crs": self.crs,
            "crs_declared": self.crs_declared,
            "lines": [line.to_dict() for line in self.lines],
            "n_active": self.n_active,
            "committed_version_id": self.committed_version_id,
            "committed_content_hash": self.committed_content_hash,
            "content_hash": self.content_hash,
            "maturity": self.maturity,
            "staleness": self.staleness,
            "staleness_detail": self.staleness_detail,
            "source": self.source,
            "validity": self.validity,
        }


def _line_summary(line: Any) -> ConstraintLineSummary:
    props = dict(getattr(line, "properties", None) or {})
    geo_kind = str(props.get("constraint_kind") or "")
    if constraint_kind_from_value(geo_kind) is None:
        geo_kind = ""
    strength_declared = "strength" in props
    try:
        strength = float(props.get("strength", 1.0)) if strength_declared else 1.0
    except (TypeError, ValueError):
        strength, strength_declared = 1.0, False
    confidence = str(props.get("confidence", "") or "unknown")
    if confidence not in ("low", "medium", "high"):
        confidence = "unknown"  # 未声明不猜
    coords = getattr(line, "coordinates", None) or []
    return ConstraintLineSummary(
        line_id=str(getattr(line, "id", "") or ""),
        name=str(getattr(line, "name", "") or ""),
        role=str(getattr(line, "role", "") or ""),
        geo_kind=geo_kind,
        active=bool(getattr(line, "active", True)),
        n_points=len(coords),
        strength=strength,
        strength_declared=strength_declared,
        confidence=confidence,
        content_fingerprint=str(props.get("content_fingerprint", "") or ""),
    )


def constraint_product_for_group(
    document: Any,
    group: Any,
    *,
    catalog: Any = None,
) -> ConstraintProduct:
    """文档约束组 → ConstraintProduct（只读投影）。"""
    from paleo_workbench.workflow.constraint_versions import (
        constraint_group_content_hash,
    )

    lines = tuple(_line_summary(line) for line in (group.lines or []))
    kinds: list[str] = []
    engine: list[str] = []
    for line in lines:
        if line.geo_kind and line.geo_kind not in kinds:
            kinds.append(line.geo_kind)
        engine_kind = to_engine_kind(line.geo_kind) if line.geo_kind else None
        if line.role:
            mapped = _ROLE_TO_ENGINE_KIND.get(line.role)
            if mapped is not None and mapped.value not in engine:
                engine.append(mapped.value)
        if engine_kind is not None and engine_kind.value not in engine:
            engine.append(engine_kind.value)

    committed_version = ""
    committed_hash = ""
    try:
        from paleo_workbench.workflow.constraint_versions import (
            current_constraint_version,
        )

        if catalog is not None:
            latest = current_constraint_version(catalog, str(group.id))
            if latest is not None:
                committed_version = str(getattr(latest, "id", "") or "")
                meta = dict(getattr(latest, "metadata", None) or {})
                committed_hash = str(meta.get("content_hash", "") or "")
    except Exception:  # noqa: BLE001 — 版本链缺失诚实为 draft
        committed_version, committed_hash = "", ""

    content_hash = ""
    try:
        content_hash, _n = constraint_group_content_hash(group)
    except Exception:  # noqa: BLE001 — 内容哈希失败 → 空串（不阻断投影）
        content_hash = ""

    # 组级 staleness：最新提交是否仍为 live 内容（resolve_constraint_ref
    # 的诚实词汇：current/stale/superseded/unknown；未提交→""=未评估）。
    staleness = ""
    staleness_detail = ""
    if committed_version:
        try:
            from paleo_workbench.workflow.constraint_versions import (
                resolve_constraint_ref,
            )

            verdict = resolve_constraint_ref(
                document, catalog,
                f"constraints:{group.id}:{committed_version}")
            staleness = str(verdict.get("status", "") or "")
            staleness_detail = str(verdict.get("detail", "") or "")
        except Exception:  # noqa: BLE001 — 解析失败 → 未评估
            staleness, staleness_detail = "unknown", "约束版本解析失败"
    elif content_hash:
        staleness = "uncommitted"
        staleness_detail = "约束组从未提交——无版本链，内容以 live 文档为准"

    validity = ""
    if committed_hash and content_hash and committed_hash != content_hash:
        validity = "存在未提交编辑（live 内容 ≠ 最新提交）"

    crs = str(getattr(group, "crs", "") or "")
    return ConstraintProduct(
        constraint_id=str(group.id),
        name=str(getattr(group, "name", "") or ""),
        target_horizon=str(getattr(group, "target_horizon", "") or ""),
        kinds=tuple(kinds),
        engine_kinds=tuple(engine),
        crs=crs,
        crs_declared=bool(crs),
        lines=lines,
        n_active=sum(1 for line in lines if line.active),
        committed_version_id=committed_version,
        committed_content_hash=committed_hash,
        content_hash=content_hash,
        maturity="committed" if committed_version else "draft",
        staleness=staleness,
        staleness_detail=staleness_detail,
        validity=validity,
    )


def constraint_products_for_document(
    document: Any,
    *,
    catalog: Any = None,
) -> list[ConstraintProduct]:
    return [
        constraint_product_for_group(document, group, catalog=catalog)
        for group in (getattr(document, "constraint_layers", None) or [])
    ]
